### ISSUE TITLE
Question about the training strategy with frozen generative LLM

### ISSUE THREAD

**Issue by lskeeper** [YeFD/RRAG#1]
When the generative LLM is frozen, how do we make sure the output embedding of the R^2-former is meaningful to the generative LLM? In that case, do you actually replace the `<R>` placeholder with the actual score of the query-doc matching model? It's not quite clear from the original paper.

**Comment by yefd** [comment]
Thanks for your questions!
1. We utilize a joint training strategy to ensure the alignment between the R^2-Former and LLM.
2. The R^2-Former outputs an embedding, not a query-doc matching score. 
3. In our implementation, we mark the `<R>` placeholder with a specific token id (such as `0` in LLaMA). We then replace its corresponding embedding with R^2-Former's output([modeling_rrag.py#L133-L135](https://github.com/YeFD/RRAG/blob/main/RRAG/models/modeling_rrag.py#L133-L135)).

**Comment by lskeeper** [comment]
Thanks for the prompt reply! But in your paper, you did mention there is an option to freeze the LLM. How was it implemented??

<img width="400" alt="截屏2024-11-04 16 43 48" src="https://github.com/user-attachments/assets/868c60f5-73c8-49d0-92ba-8a753493c111">
<img width="420" alt="截屏2024-11-04 16 43 55" src="https://github.com/user-attachments/assets/1e85ea3a-d2f3-402e-beec-3d508d01c54f">

**Comment by yefd** [comment]
Yes, the `freeze_llm` parameter control to freeze the LLM or not. [modeling_rrag.py#L81-L84](https://github.com/YeFD/RRAG/blob/main/RRAG/models/modeling_rrag.py#L81-L84) shows the trigger.
Conversely, you can fine-tune the LLM by removing `--freeze_llm` in the script. Optionally, add `--use_lora` to enable PEFT (such as [train_nq_lora.sh .sh](https://github.com/YeFD/RRAG/blob/main/scripts/train/train_nq_lora.sh)).

**Comment by lskeeper** [comment]
> Yes, the `freeze_llm` parameter control to freeze the LLM or not. [modeling_rrag.py#L81-L84](https://github.com/YeFD/RRAG/blob/main/RRAG/models/modeling_rrag.py#L81-L84) shows the trigger. Conversely, you can fine-tune the LLM by removing `--freeze_llm` in the script. Optionally, add `--use_lora` to enable PEFT (such as [train_nq_lora.sh .sh](https://github.com/YeFD/RRAG/blob/main/scripts/train/train_nq_lora.sh)).

I meant, when you freeze the LLM, do you have a special strategy to deal with the `<R>` token? If R^2-Former is not aligned with the LLM, will it cause unexpected/undefined outcome?

**Comment by yefd** [comment]
The `<R>` token is just a placeholder, and it will be replaced before LLM processes the input embeddings. So, the LLM takes several embeddings as additional inputs. And, the training strategy ensures the additional embeddings from R^2-Former are meaningful for the LLM. 
If the R^2-Former is not aligned, the LLM may be disturbed by these additional embeddings and generate some meaningless tokens.

**Comment by lskeeper** [comment]
Thanks again for the reply!

> If the R^2-Former is not aligned, the LLM may be disturbed by these additional embeddings and generate some meaningless tokens

That's my understanding as well. But from the paper, I don't see how the R^2-Former model is aligned with LLM if the LLM is frozen (since the R^2-Former model is only trained on the QDM task). So I'm a bit confused why this works when LLM is frozen. Hope that clarifies my question :)

**Comment by yefd** [comment]
Oh, I got your point. 
The key is the joint training, which involves LM and QDM tasks(not only QDM). The LM task makes the R^2-Former can receive feedback from the LLM(even frozen).
It is a common training strategy in multi-modality alignment, such as blip-2. You can regard the additional embeddings as another modal information for better understanding.

**Comment by lskeeper** [comment]
ah, I think I understand now. So when the LLM is frozen, I thought the LM loss is also removed. It makes sense to me now. Thanks a lot for the explanation!
